MarkBind/markbind
# Fix custom components markdown parsing

**What is the purpose of this pull request?**


Resolves #958. 

This issue is also related to #1534. We were prompted to re-visit #958 as it needs to be fixed before we can proceed to implement #1534. 

As @ang-zeyu mentioned, we have never defined how our components should behave as html tags (block / inline (default)) regarding markdown. 

Thus, we decided to adapt VuePress implementation to fit our own needs. VuePress treats every custom component (or unknown tags) as block-level element. This applies mostly in our use-case as well, except for `<panel minimized>` component (panels which are minimized), which is treated as inline element. 



**Overview of changes:**
Adapted VuePress `markdown-it` [patch](https://github.com/vuepress/vuepress-next/tree/main/packages/%40vuepress/markdown/src/plugins/customComponentPlugins) for custom components. 

**Anything you'd like to highlight / discuss:**
The patch used by VuePress replaces the `html_block` rule in `markdown-it`. However, on our side, we are also replacing `html_block` rule with `special-escape-tags` already. 

Thus, instead of modifying / merging `special-escape-tags` rule and VuePress's `custom-component` rule, I decided to just add VuePress's custom component patch rule under `special-escape-tags` rule. 

Should I look into combining the rules together somehow? Seeing that there is a bit of code duplication here and time-wise it is a little more expensive as `markdown-it` has to run through an additional rule (which can be combined into one).

But I think I am more comfortable with not combining them for separation of concern reasons. It would be easier to read the code with them being separate in my opinion. 

**Testing instructions:**
`npm run test`

Try out the examples as raised by @ang-zeyu in #958. The bugs should be fixed. 

However, when you use the minimized option for panel, then the "bug" (not actually a bug since it is expected behaviour) should still be present, as minimized panel is treated as an inline element instead of block-level element.



**Proposed commit message: (wrap lines at 72 characters)**
Fix custom components markdown parsing.

We have never defined the behavior of our custom components as html tags
(whether block-level or inline) in markdown parsing. This causes some 
undesirable edge cases during markdown parsing, as raised up in #958. 

Let's adopt the way VuePress define the behaviour for custom components
and adapt it to fit our needs. 



---

**Checklist:** :ballot_box_with_check:
